transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,874 @@
+            .map_err(Error::IoError)
+            .map_err(TransportError::Other)?;
+
+        let listen_addr = socket
+            .local_addr()
+            .map_err(Error::IoError)
+            .map_err(TransportError::Other)?;
+        debug!("listening on {}", listen_addr);
+
+        let udp_mux = UDPMuxNewAddr::new(UDPMuxParams::new(socket));
+
+        Ok(WebRTCListenStream::new(
+            listener_id,
+            listen_addr,
+            self.config.clone(),
+            udp_mux,
+            self.id_keys.clone(),
+        ))
+    }
+}
+
+impl Transport for WebRTCTransport {
+    type Output = (PeerId, Connection);
+    type Error = Error;
+    type ListenerUpgrade = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        let id = ListenerId::new();
+        let listener = self.do_listen(id, addr)?;
+        self.listeners.push(listener);
+        Ok(id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Poll all listeners.
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+        match self.listeners.poll_next_unpin(cx) {
+            Poll::Ready(Some(ev)) => Poll::Ready(ev),
+            _ => Poll::Pending,
+        }
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let config = self.config.clone();
+        let our_fingerprint = self.cert_fingerprint();
+        let id_keys = self.id_keys.clone();
+
+        let sock_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if sock_addr.port() == 0 || sock_addr.ip().is_unspecified() {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+        let remote = addr.clone(); // used for logging
+        trace!("dialing address: {:?}", remote);
+
+        let udp_mux = if let Some(l) = self.listeners.iter_mut().next() {
+            l.udp_mux.clone()
+        } else {
+            return Err(TransportError::Other(Error::NoListeners));
+        };
+        let se = build_setting_engine(udp_mux, &sock_addr, &our_fingerprint);
+        let api = APIBuilder::new().with_setting_engine(se).build();
+
+        // [`Transport::dial`] should do no work unless the returned [`Future`] is polled. Thus
+        // do the `set_remote_description` call within the [`Future`].
+        Ok(async move {
+            let peer_connection = api
+                .new_peer_connection(config)
+                .map_err(Error::WebRTC)
```

Comment:
The `Error` implements `From` so you should be able to skip all of these `map_err` calls.